File: Programs/matlab/I-vectors/eval_by_bosaris.m
function [eer, mindcf12, actdcf12, calw, pavfunc] = ...
    eval_by_bosaris(sre12_evlfile, sre12_keyfile, fast_keyloading, calw, pavfunc)
% Use Bosaris toolkit for evaluation. 
% be
% Input: 
%   sre12_evlfile      - Evaluation file that match the segment name of key file
%   keyfile            - Optional key file (work for SRE12 key file)
%   fast_keyloading    - Optinoal flag for fast keyfile loading (load a pre-stored .mat key file)
%   calw               - Calibration weights in 1x2 array
% Output:
%   eer                - Equal error rate
%   dcf12              - Minimum DCF as defined in SRE12
%   res12              - Result12 object in Bosaris toolkit
% Example:
%   evl2evl('evl/fw60/gplda60_male_cc4_1024c.evl', 'ndx/male/core-core_8k_male_cc4.ndx','evl/fw60/sre12_gplda60_male_cc4_1024c.evl');
%   [eer,dcf12,res12] = eval_by_bosaris('evl/fw60/sre12_gplda60_male_cc4_1024c.evl','../../key/NIST_SRE12_core_trial_key.v1',true);

% Setup environment for Bosaris
%bosaris_base = '~/so/Matlab/bosaris';
%addpath(genpath(bosaris_base));

% Note that user need to run scripts/re_arrange_evl.pl or evl2evl.m to produce an .evl file that can match the
% segment name of the key file.


% Set to true if .mat format of the key file has been saved from previous run
if ~exist('sre12_keyfile','var'),
    sre12_keyfile = '../../key/NIST_SRE12_core_trial_key.v1';          
end
if ~exist('fast_keyloading','var'),
    fast_keyloading = 'true';
end

% Setup Bosaris toolkit
%run('~/so/Matlab/bosaris/setup_bosaris.m');

% Alternatively, you may call the following private function
setup_bosaris('./bosaris_toolkit');       % Assuming ./bosaris_tk contains the toolkit

% Load key. If .mat file for key has been saved, it is quicker to load the .mat file
if (fast_keyloading == true),
    fprintf('Loading SRE12 key file %s\n',strcat(sre12_keyfile,'.mat'));
    key12 = Key.read(strcat(sre12_keyfile,'.mat')); 
else    
    fprintf('Loading SRE12 key file %s\n',strcat(sre12_keyfile,'.csv'));
    key12 = Key.read(strcat(sre12_keyfile,'.csv'));
    % Save SRE12 key to .mat file for improving loading speed
    fprintf('Saving SRE12 key as %s\n',strcat(sre12_keyfile,'.mat'));
    save_mat(key12,strcat(sre12_keyfile,'.mat'));
end

% Load SRE12 .evl file
fprintf('Loading SRE12 evl file %s\n',sre12_evlfile);
scr12 = Scores.read(sre12_evlfile);

% filter the key so that it contains only trials for which we have scores.
key12 = key12.filter(scr12.modelset,scr12.segset,true);

% Get normalized min DCF
res12 = Results12(scr12,key12);
prior1 = effective_prior(0.01,1,1);      % P_tgt_A1 in SRE12
prior2 = effective_prior(0.001,1,1);     % P_tgt_A2 in SRE12
minNormDcf12 = res12.get_norm_min_dcf(prior1,prior2);
actNormDcf12 = res12.get_norm_act_dcf(prior1,prior2);
Cllr = cllr(res12.tar,res12.non);
minCllr = min_cllr(res12.tar,res12.non);

% Print results based on uncalibrated scores
fprintf('Before score calibration\n');
fprintf('EER=%.2f; minNormDcf=%.3f; actNormDcf=%.3f; minCllr=%.3f; Cllr=%.3f\n',...
        res12.eer*100,minNormDcf12,actNormDcf12,minCllr,Cllr);

% Compute linear calibration weights and PAV function if they are not provided.    
if ~exist('calw','var'),
    [~,~,calw] = train_linear_calibration(res12.tar,res12.non,prior1,[],100,true);
end
if ~exist('pavfunc','var'),
    pavfunc = pav_calibration(res12.tar,res12.non,0.001);
end

% Perform linear score calibration
scr12_c = scr12;
scr12_c.scoremat = calw(1)*scr12.scoremat + calw(2);  % calw = [scale offset]
res12_c = Results12(scr12_c,key12);
minNormDcf12 = res12_c.get_norm_min_dcf(prior1,prior2);
actNormDcf12 = res12_c.get_norm_act_dcf(prior1,prior2);
Cllr = cllr(res12_c.tar,res12_c.non);
minCllr = min_cllr(res12_c.tar,res12_c.non);

% Print results based on calibrated scores
fprintf('After linear score calibration\n');
fprintf('EER=%.2f; minNormDcf=%.3f; actNormDcf=%.3f; minCllr=%.3f; Cllr=%.3f\n',...
        res12.eer*100,minNormDcf12,actNormDcf12,minCllr,Cllr);

% Perform PAV score calibration
scr12_c = scr12;
s = pavfunc(scr12.scoremat(:));
scr12_c.scoremat = reshape(s, size(scr12.scoremat));
res12_c = Results12(scr12_c,key12);
minNormDcf12 = res12_c.get_norm_min_dcf(prior1,prior2);
actNormDcf12 = res12_c.get_norm_act_dcf(prior1,prior2);
Cllr = cllr(res12_c.tar,res12_c.non);
minCllr = min_cllr(res12_c.tar,res12_c.non);

% Print results based on calibrated scores
fprintf('After PAV score calibration\n');
fprintf('EER=%.2f; minNormDcf=%.3f; actNormDcf=%.3f; minCllr=%.3f; Cllr=%.3f\n',...
        res12.eer*100,minNormDcf12,actNormDcf12,minCllr,Cllr);

eer = res12.eer*100;
mindcf12 = minNormDcf12;
actdcf12 = actNormDcf12;

return;
%%

function setup_bosaris(bosaris_base)
% Setup environment for Bosaris
addpath(genpath(bosaris_base));


%--------- FOR CALIBRATION OF THRESHOLD ONLY -------------------

% % Perform calibration. calw is the calibration weights [scale, offset]
% [~, calw] = linear_calibrate_scores(scr12,key12,[],100,prior1,true);
% scr12_c = scr12;
% scr12_c.scoremat = calw(1)*scr12.scoremat + calw(2);  % calw = [scale offset]
% res12_c = Results12(scr12_c,key12);
% minNormDcf12 = res12_c.get_norm_min_dcf(prior1,prior2);
% actNormDcf12 = res12_c.get_norm_act_dcf(prior1,prior2);
% 
% % Print results based on calibrated scores
% fprintf('After score calibration\n');
% fprintf('EER=%.2f; minNormDcf=%.3f; actNormDcf=%.3f\n',res12_c.eer*100,minNormDcf12,actNormDcf12);
% 
% % Compute compounded LLR
% load('mat/tgt-tst-scoremat.mat');       % Load scoremat (including trials not specified in SRE protocol)
% scr12_c = scr12;
% scr12_c.scoremat = calw(1)*scoremat + calw(2);   % Calibrate scores
% N = length(scr12_c.modelset);
% %scr12_cllr = llrTrans_simple2compound(scr12_c.scoremat,[ones(1,N)/N,1]/2);
% scr12_cllr = scr12_c.compoundLLR([ones(1,N)/N,1]/2);
% res12_cllr = Results12(scr12_cllr,key12);
% minNormDcf_cllr = res12_cllr.get_norm_min_dcf(prior1,prior2);
% actNormDcf_cllr = res12_cllr.get_norm_act_dcf(prior1,prior2);
% 
% % Print results based on calibrated compound LLR scores
% fprintf('Performance of compound LLR\n');
% fprintf('EER=%.2f; minNormDcf=%.3f; actNormDcf=%.3f\n',...
%     res12_cllr.eer*100,minNormDcf_cllr,actNormDcf_cllr);
% 
% % Use linear_calibrate_scores_dev_eval.m to perform score calibration
% 
% break
% 
% %%
% 
% sre10_keyfile = '../../../nist10/key/coreext-coreext.trialkey.mat';
% sre10_evlfile = '../../evl/fw60/sre10_gplda60_male_xphonecall-tel_1024c.evl';
% sre12_ndxfile = '../../ndx/male/core-core_8k_female_cc4.ndx';
% 
% % Convert SRE10 text key file to .mat file
% %key10 = Key.read('../../../nist10/key/coreext-coreext.trialkey.csv');
% %save_mat(key10,'../../../nist10/key/coreext-coreext.trialkey.mat');
% 
% fprintf('Loading SRE10 key file %s\n',sre10_keyfile);
% key10 = Key.read(sre10_keyfile);
% 
% %====================================================================
% % Use SRE10 key and scores for training the calibration parameters
% % Note that the scores are based on the PLDA model and T matrix trained 
% % for SRE12 so that the scores and i-vec are competable with those in SRE12
% %====================================================================
% 
% % Compute suff stats for targets and test segments
% comp_suf_stats('ubm/fw60/bg.mix.male.cmsY.1024c','../../../nist10/matlab/jfa/lst/fw60/male_xtarget_stats.lst','../../../nist10/matlab/jfa/stats/fw60/male_xtarget_mix_stats_1024c.mat',4);
% comp_suf_stats('ubm/fw60/bg.mix.male.cmsY.1024c','../../../nist10/matlab/jfa/lst/fw60/male_xtest_phonecall_tel_stats.lst','../../../nist10/matlab/jfa/stats/fw60/male_xtest-phonecall-tel_mix_stats_1024c.mat',4);
% 
% % Compute i-vectors for targets and test segments 
% comp_ivecs('ubm/fw60/bg.mix.male.cmsY.1024c','mat/fw60/male_mix_t500_1024c.mat','../../../nist10/matlab/jfa/stats/fw60/male_xtarget_mix_stats_1024c.mat','../../../nist10/matlab/jfa/mat/fw60/male_xtarget_mix_t500_w_1024c.mat',4);
% comp_ivecs('ubm/fw60/bg.mix.male.cmsY.1024c','mat/fw60/male_mix_t500_1024c.mat','../../../nist10/matlab/jfa/stats/fw60/male_xtest-phonecall-tel_mix_stats_1024c.mat','../../../nist10/matlab/jfa/mat/fw60/male_xtest-phonecall-tel_mix_t500_w_1024c.mat',4);
% 
% % Compute the scores of SRE10 and save it to .evl file (in SRE12 format)
% score_gplda_w({'../../../nist10/matlab/jfa/mat/fw60/male_xtarget_mix_t500_w_1024c.mat','../../../nist10/matlab/jfa/mat/fw60/male_xtarget_mix_t500_w_1024c.mat'},'','../../../nist10/matlab/jfa/mat/fw60/male_xtest-phonecall-tel_mix_t500_w_1024c.mat','../../../nist10/matlab/jfa/lst/fw60/male_xndx_phonecall_tel_stats.lst','','','None',sre10_evlfile,'mat/fw60/male_mix_t500_gplda_1024c.mat');
% 
% % Load SRE10 .evl file phonecall-tel
% fprintf('Loading SRE12 evl file %s\n',sre10_evlfile);
% scr10 = Scores.read(sre10_evlfile);
% 
% % Load SRE12 .ndx file for calibration
% fprintf('Loading SRE12 ndx file %s\n',sre10_ndxfile);
% ndx12 = Ndx.read(sre12_ndxfile);
% 
% % Use SRE10 to train a score calibrator and then calibrate the scores of SRE12
% [~,~,w_lin] = linear_calibrate_scores_dev_eval(scr10,scr12,key10,ndx12,[],10,prior1,true);

    

File: Programs/python/mPLDA/eval_by_bosaris_.py
import h5py as h5
from scipy.io import loadmat, savemat
import numpy as np
# import cupy as np
from utili_ import dict2h5
from bosaris_toolkit.key import Key
from bosaris_toolkit.scores import Scores
from bosaris_toolkit.detplot import effective_prior
from bosaris_toolkit.detplot import rocch
from bosaris_toolkit.detplot import rocch2eer
from bosaris_toolkit.detplot import sigmoid

class Results:

    def __init__(self):
        self.tar = np.empty(0, dtype=np.float64)
        self.non = np.empty(0, dtype=np.float64)
        self.knnon = np.empty(0, dtype=np.float64)
        self.uknnon = np.empty(0, dtype=np.float64)
        self.Pmiss = np.empty(0, dtype=np.float64)
        self.Pfa = np.empty(0, dtype=np.float64)
        self.Pfa_kn = np.empty(0, dtype=np.float64)
        self.Pfa_ukn = np.empty(0, dtype=np.float64)
        self.thresholds = np.empty(0, dtype=np.float64)


def full_roc(tar_scores, nontar_scores, knnontar_scores, uknnontar_scores, nThresholds=20000):
    """
     Compute the full ROC by sweeping nThresholds ranging from
     the smallest score and the largest scores in tar_scores, and nontar_scores
     Author: M.W. Mak
    """

    score = np.hstack((tar_scores.T, nontar_scores.T))
    dx = (np.max(score) - np.min(score)) / nThresholds
    assert dx > 0, 'The two score distributions do not overlap\n'
    maxs = np.max(score) - dx
    mins = np.min(score) + dx
    X = np.array(range(nThresholds) * (maxs - mins) / (nThresholds - 1) + mins)

    # Compute FP and FN at different thresholds
    fn = np.zeros(shape=(len(X),), dtype=np.float64)
    fp = np.zeros(shape=(len(X),), dtype=np.float64)
    for i in range(len(X)):
        fn[i] = len(np.where(tar_scores <= X[i])[0])
        fp[i] = len(np.where(nontar_scores > X[i])[0])

    # Compute FAR and FRR at different thresholds
    Pmiss = fn / len(tar_scores)
    Pfa = fp / len(nontar_scores)

    for i in range(len(Pfa)):
        if Pmiss[i] >= Pfa[i]:
            break
    eer = (Pmiss[i] + Pfa[i]) / 2

    # Compute the Pfa of known and unknown nontargets
    for i in range(len(X)):
        fp[i] = len(np.where(knnontar_scores > X[i])[0])
    Pfa_kn = fp / len(knnontar_scores)
    for i in range(len(X)):
        fp[i] = len(np.where(uknnontar_scores > X[i])[0])

    # To prevent the case where there is no unknown non-target (e.g., CC4 male))
    if uknnontar_scores.size == 0:
        Pfa_ukn = fp
    else:
        Pfa_ukn = fp / len(uknnontar_scores)

    return Pmiss, Pfa, Pfa_kn, Pfa_ukn, eer, X

def get_norm_min_dcf(res, prior1, prior2):

    Pknown = 0.5
    Ptar1 = prior1
    Pnon1 = 1 - prior1
    Ptar2 = prior2
    Pnon2 = 1 - prior2
    res.Pmiss, res.Pfa, res.Pfa_kn, res.Pfa_ukn, _, res.thresholds = full_roc(res.tar, res.non, res.knnon, res.uknnon)
    cdet1 = np.dot(np.hstack((Ptar1, Pnon1)), np.vstack((res.Pmiss.T, Pknown * res.Pfa_kn.T + (1 - Pknown) * res.Pfa_ukn.T)))
    cdet2 = np.dot(np.hstack((Ptar2, Pnon2)), np.vstack((res.Pmiss.T, Pknown * res.Pfa_kn.T + (1 - Pknown) * res.Pfa_ukn.T)))
    ncdet1 = cdet1 / np.min((Ptar1, Pnon1))
    ncdet2 = cdet2 / np.min((Ptar2, Pnon2))
    mindcf = (np.min((np.min(ncdet1), 2)) + np.min((np.min(ncdet2), 2))) / 2
    return mindcf

def fast_sre12_actDCF(tar, knnon, uknnon, plo, normalize = False):
    """
         Modified by Mak based on fast_actDCF.m for SRE12 actual DCF
     Computes the actual average cost of making Bayes decisions with scores
     calibrated to act as log-likelihood-ratios. The average cost (DCF) is
     computed for a given range of target priors and for unity cost of error.
     If un-normalized, DCF is just the Bayes error-rate.

      Usage examples:  dcf = fast_actDCF(tar,non,-10:0.01:0)
                       norm_dcf = fast_actDCF(tar,non,-10:0.01:0,true)
                       [dcf,pmiss,pfa] = fast_actDCF(tar,non,-10:0.01:0)

      Inputs:
        tar: a vector of T calibrated target scores
        non: a vector of N calibrated non-target scores
             Both are assumed to be of the form

                   log P(data | target)
           llr = -----------------------
                 log P(data | non-target)

             where log is the natural logarithm.
        knnon: a vector of N_kn calibrated known-nontarget scores
        uknnon: a vector of N_ukn calibrated unknown-nontarget scores

        plo1 and plo2: an ascending vector of log-prior-odds, plo = logit(Ptar)
                                                        = log(Ptar) - log(1-Ptar)

        normalize: (optional, default false) return normalized dcf if true.


       Outputs:
         dcf: a vector of DCF values, one for every value of plo.

                 dcf(plo) = Ptar(plo)*Pmiss(plo) + (1-Ptar(plo1))*Pfa(plo)

              where Ptar(plo) = sigmoid(plo) = 1./(1+exp(-plo)) and
              where Pmiss and Pfa are computed by counting miss and false-alarm
              rates, when comparing 'tar' and 'non' scores to the Bayes decision
              threshold, which is just -plo. If 'normalize' is true, then dcf is
              normalized by dividing by min(Ptar,1-Ptar).

          Pmiss: empirical actual miss rate, one value per element of plo.
                 Pmiss is not altered by parameter 'normalize'.

          Pfa: empirical actual false-alarm rate, one value per element of plo.
               Pfa is not altered by parameter 'normalize'.

     Note, the decision rule applied here is to accept if

        llr >= Bayes threshold.

     or reject otherwise. The >= is a consequence of the stability of the
     sort algorithm , where equal values remain in the original order.
    """

    assert len(tar.shape) <= 1
    assert len(knnon.shape) <= 1
    assert len(uknnon.shape) <= 1
    assert len(plo.shape) <= 1

    assert (np.array([plo]) == np.sort(np.array([plo]))).any(), 'Parameter plo1 must be in ascending order.'

    D = plo.size
    T = tar.size
    N_kn = knnon.size
    N_ukn = uknnon.size

    tar = tar.reshape((1, -1))
    knnon = knnon.reshape((1, -1))
    plo = plo.reshape((1, -1))
    uknnon = uknnon.reshape((1, -1))

    ii = np.argsort(np.hstack((-plo, tar)))[0]  # -plo are thresholds
    r = np.zeros(shape=(T + D), dtype=np.float64)
    r[ii] = np.array(range(1, T+D+1))
    r = r[0:D]  # rank of thresholds
    Pmiss = r - np.array(range(D, 0, -1))

    ii = np.argsort(np.hstack((-plo, knnon)))[0]  # -plo are thresholds
    r = np.zeros(shape=(N_kn+D), dtype=np.float64)
    r[ii] = np.array(range(1, N_kn+D+1))
    r = r[0:D]   # rank of thresholds
    Pfa_kn = N_kn - r + np.array(range(D, 0, -1))

    ii = np.argsort(np.hstack((-plo, uknnon)))[0]  # -plo are thresholds
    r = np.zeros(shape=(N_ukn+D), dtype=np.float64)
    r[ii] = np.array(range(1, N_ukn+D+1))
    r = r[0:D]   # rank of thresholds
    Pfa_ukn = N_ukn - r + np.array(range(D, 0, -1))

    Pmiss = Pmiss / T
    Pfa_kn = Pfa_kn / N_kn

    if N_ukn > 0:   # Prevent divided by 0 in case there is on unknown non-targets
        Pfa_ukn = Pfa_ukn / N_ukn

    Ptar = sigmoid(plo)
    Pnon = sigmoid(-plo)
    Pknown = 0.5

    dcf = Ptar * Pmiss + Pnon * (np.dot(Pknown, Pfa_kn) + np.dot((1 - Pknown), Pfa_ukn))

    if normalize:
        dcf = dcf / np.min((Ptar, Pnon))

    return dcf[0], Pmiss, Pfa_kn, Pfa_ukn

def logit(temp):
    return np.log(temp / (1 - temp))

def get_norm_act_dcf(res, prior1, prior2):
    actdcf1, _, _, _ = fast_sre12_actDCF(res.tar, res.knnon, res.uknnon, logit(prior1), normalize=True)
    actdcf2, _, _, _ = fast_sre12_actDCF(res.tar, res.knnon, res.uknnon, logit(prior2), normalize=True)
    actdcf = (actdcf1 + actdcf2) / 2
    return actdcf

def eval_by_bosaris(sre12_evlfile, sre12_keyfile = './key/NIST_SRE12_core_trial_key.v1', fast_keyloading = True):
    """
     Use Bosaris toolkit for evaluation.
     be
     Input:
       sre12_evlfile      - Evaluation file that match the segment name of key file
       keyfile            - Optional key file (work for SRE12 key file)
       fast_keyloading    - Optinoal flag for fast keyfile loading (load a pre-stored .h5 key file)
     Output:
       eer                - Equal error rate
       dcf12              - Minimum DCF as defined in SRE12
       res12              - Result12 object in Bosaris toolkit
     Example:
       evl2evl('evl/fw60/gplda60_male_cc4_1024c.evl', 'ndx/male/core-core_8k_male_cc4.ndx','evl/fw60/sre12_gplda60_male_cc4_1024c.evl');
       [eer,dcf12,res12] = eval_by_bosaris('evl/fw60/sre12_gplda60_male_cc4_1024c.evl','../../key/NIST_SRE12_core_trial_key.v1',true);

     Note that user need to run scripts/re_arrange_evl.pl or evl2evl_.py to produce an .evl file that can match the
     segment name of the key file.

     Set to true if .h5 format of the key file has been saved from previous run

    """

    sre12_keyfile_h5_path = sre12_keyfile + '.h5'

    if not fast_keyloading:

        temp = loadmat(sre12_keyfile + '.mat')['key'][0][0]
        key12 = {'modelset': np.array([item[0][0].encode() for item in temp[0]]),\
                 'segset': np.array([item[0][0].encode()for item in temp[1]]),\
                 'tar': temp[2],\
                 'non': temp[3],\
                 'knnon': temp[4], \
                 'uknnon': temp[5]}

        dict2h5(key12, file=sre12_keyfile_h5_path)

    key12 = Key(key_file_name=sre12_keyfile_h5_path)   # Only read from .h5 file

    scr12 = Scores()
    scr12 = scr12.read_evl(sre12_evlfile)

    # filter the key so that it contains only trials for which we have scores.
    key12 = key12.filter(scr12.modelset, scr12.segset, keep=True)

    prior1 = effective_prior(0.01, 1, 1)  # P_tgt_A1 in SRE12
    prior2 = effective_prior(0.001, 1, 1)  # P_tgt_A2 in SRE12

    res12 = Results()
    res12.tar, res12.non, res12.knnon, res12.uknnon = scr12.get_tar_non(key12)

    res12.Pmiss, res12.Pfa = rocch(res12.tar, res12.non)
    res12.eer = rocch2eer(res12.Pmiss, res12.Pfa)

    minNormDcf12 = get_norm_min_dcf(res12, prior1, prior2)
    actNormDcf12 = get_norm_act_dcf(res12, prior1, prior2)

    # Print results based on uncalibrated scores
    print('Before score calibration\n')
    print('EER=%.2f; minNormDcf=%.3f; actNormDcf=%.3f\n' % (res12.eer*100, minNormDcf12, actNormDcf12))

    eer = res12.eer * 100
    dcf12 = minNormDcf12

    return eer, dcf12, res12